documents/final-submission-document/chapters/4_implementation.tex

```diff
@@ -61,15 +61,87 @@ \subsection{Limitations of \emph{\fullref{sec:requirement-audit-trail}}}
+With an assumed count of approximately 1000 \glspl{topic}, this would result in a great performance overhead. Therefore, we decided to not implement this requirement.
+
+\subsubsection{Filtering and Paging of \glspl{persisted-item}}
+
+Requirements~\fullref{sec:requirement-filter-persisted-items} and \fullref{sec:paged-persisted-items} were built on the premise that filtering and paging of \glspl{persisted-item} would be possible with the \gls{jabber-search} \cite{xep-0055} and \gls{result-set-management} \cite{xep-0059} \glspl{xep}.
+
+Sadly, the \gls{publish-subscribe} \gls{xep} does not explicitly define an integration with \gls{jabber-search}.
+The main reason for this is probably the exchangeable \glspl{persisted-item} content format.
+
+Retrieving multiple \glspl{persisted-item} in \gls{result-set-management} pages was added in version 1.12 (2008-09-03) only of the \gls{publish-subscribe} \gls{xep}.
+A \gls{xmpp} server does not report, which version of the standard it supports.
+
+Therefore, we could not presume an implementation of \gls{result-set-management}.
+In fact, the openfire \gls{xmpp} server we used in our setup has no support for retrieving \glspl{persisted-item} with \gls{result-set-management}.
+
+\subsubsection{Create and Configure \glspl{topic}}
+
+We have four requirements that are concerned with the initial configuration of \glspl{topic}:
+\begin{itemize}
+  \item \fullref{sec:requirement-topic-default-configuration}
+  \item \fullref{sec:requirement-collection-default-configuration}
+  \item \fullref{sec:requirement-initial-topic-consumer-provider}
+  \item \fullref{sec:requirement-initial-collection-consumer}
+\end{itemize}
+
+The implementation of a initial configuration of \glspl{topic} was not possible due to limitations in the \gls{xep} \gls{data-forms} and \gls{publish-subscribe} standards. According to these standards, the form to configure \glspl{topic} can only be requested after the nodes have been created. It is not possible to request a general configuration form prior to creating the according \gls{topic}.~\cite{xep-0004, xep-0060}
+
+Therefore, we decided to split creation and configuration of \glspl{topic} into a two step process.
+
+\subsection{Limitations of the Openfire \gls{xmpp} Server}\label{sec:limitations-of-the-openfire-xmpp-server}
+
+As discussed in Section~\ref{sec:development-setup}, the openfire \gls{xmpp} server was used in the development setup. This section details on the encountered limitations while implementing the \gls{xmpp-grid} \gls{broker}.
+
+\subsubsection{WebSocket SASL EXTERNAL Support}
+
+At time of this thesis, openfire does not support SASL EXTERNAL in combination with \gls{xmpp} over WebSockets.
+Therefore, the current implementation of the \gls{xmpp-grid} \gls{broker} was developed with \gls{bosh}, but also support communication over WebSockets thanks to the stanza.io \gls{xmpp} library\footnote{\url{https://github.com/legastero/stanza.io}}.
+
+\subsubsection{Lost Updates}
```

Comment:
I was not sure, if we should also talk about parallel edits here, what do you think @raphiz?

Reply:
As discussed, parallel edits will be discussed in the concurrency chapter